### DISCUSSION TITLE
Can't jump to *.c (implementation), only jump to *.h (header definition)

### DISCUSSION THREAD

**Discussion by andy1990zx** [microsoft/vscode-cpptools#9464] [Q&A] [answered]
Hi,

I took several days to solve this problem but still failed.
When I jump to a routine, it can only jump to *.h, not *.c. But if I search the same routime and open the correct *.c, next time I jump to this routine it will jump to *.c
![GIF](https://user-images.githubusercontent.com/37032134/174279274-10e4dd74-90a1-490f-92f0-dea578f98e0b.gif)

Seems only when the *.c been loaded into VS Code then it can be parsed correctly.
![image](https://user-images.githubusercontent.com/37032134/174278657-228733e1-a116-4f98-ad66-5165f806aaa7.png)

I have some questions here:
1. Is my understanding correct?
2. If Q1 = Yes, why the *.h can be parsed at the first time, without been loaded into VS Code?
3. If Q1 = No, so how to jump to *.c at the first time?
    (I tried to use C++ Intellisense, it works, but has some other issues for me. e.g. can't jump to local variable.
    If I enable both C/C++ and C++ Intellisense, it will show two jump targets, one is .h from C/C++, one is .c from C++ Intellisense.)

**Comment by sean-mcmanus** [comment, marked as answer]
Normally when you open a workspace folder, our extension should parse all the .c/.cpp files by default. If you set C_Cpp.loggingLevel to "Debug" and do a "C/C++: Reset IntelliSense Database" command you'll see a bunch of "tag parsing" logging messages. You'll want to make sure all the .c files are getting tag parsed. It's possible something may be causing them to be skipped, such as a files.exclude setting being set or you may have a configurationProvider set that isn't configured properly so it's not sending the proper browse.path settings. The logging should indicate what folders will be parsed for workspace symbols. If you're using a configurationProvider, then running the C/C++: Log Diagnostics command should also show the browse configurations being sent.

**Reply by andy1990zx** [reply]
_Normally when you open a workspace folder, our extension should parse all the .c/.cpp files by default._
That's a great news for me! After set C_Cpp.loggingLevel to "Debug" and do a "C/C++: Reset IntelliSense Database" command, it shows ZERO file discovered
_Discovering files...
  Discovering files: 0 file(s) processed_

I think that's the right direction. So I opened the parent folder of my "old" work space, it discovered all files and everything is OK!
I opened the "old" work space again, 0 file.

I have a folder layout as below:
XxxProject  ----> This is the parent folder, tag parsing works fine if I open it by VS CODE
--Code/  ----> This is my "old" workspace, all source code files are here. And 0 files discovered if I open it by VS CODE
--Doc/
--ReleaseData/
--.gitignore

I run C/C++: Log Diagnostics command on "old" workspace and patent:
FROM "old" (issue):

-------- Diagnostics - 2022/6/18 14:48:00
Version: 1.10.7
Current Configuration:
{
    "name": "Win32",
    "includePath": [
        "${workspaceFolder}/**"
    ],
    "defines": [
        "_DEBUG",
        "UNICODE",
        "_UNICODE"
    ],
    "cStandard": "c17",
    "cppStandard": "c++17",
    "intelliSenseMode": "windows-msvc-x64",
    **"configurationProvider": "ms-vscode.makefile-tools",**
    "compilerPathIsExplicit": false,
    "cStandardIsExplicit": true,
    "cppStandardIsExplicit": true,
    "intelliSenseModeIsExplicit": true,
    "macFrameworkPath": [],
    "mergeConfigurations": false,
    "compilerPath": "D:\\usr\\bin\\gcc",
    "browse": {
        "path": [
            "${workspaceFolder}/**"
        ]
    }
}
Custom browse configuration: 
{
    "browsePath": [],
    "compilerArgs": []
}


FROM parent (no issue):
-------- Diagnostics - 2022/6/18 14:49:21
Version: 1.10.7
Current Configuration:
{
    "name": "Win32",
    "includePath": [
        "${workspaceFolder}/**"
    ],
    "defines": [
        "_DEBUG",
        "UNICODE",
        "_UNICODE"
    ],
    "compilerPath": "D:\\usr\\bin\\gcc",
    "cppStandard": "c++17",
    "intelliSenseMode": "windows-gcc-x64",
    "cStandard": "c17",
    "intelliSenseModeIsExplicit": false,
    "cStandardIsExplicit": true,
    "cppStandardIsExplicit": false,
    "mergeConfigurations": false,
    "compilerPathIsExplicit": false,
    "browse": {
        "path": [
            "${workspaceFolder}/**"
        ]
    }
}
No active translation units.


So the _"configurationProvider": "ms-vscode.makefile-tools",_ is my priority, but I found it's empty in the setting:
![image](https://user-images.githubusercontent.com/37032134/174426799-6e0bb40d-626e-49d8-a4f6-1752e4fd1398.png)

Uninstall the makefile-tools extension, "configurationProvider": "ms-vscode.makefile-tools", still here.

How can I remove it?

**Reply by andy1990zx** [reply]
Finally, I used C/C++: Edit configurations... to remove the configurationProvider, it works fine now...

The VS CODE (or C/C++) design doesn't make much sense:
1. There is a global configuration page which include many C/C++ configurations, but still a dedicate C/C++ configuration page
2. Even a CTRL+SHIFT+P command window... And no UI for it
3. Seems the c_cpp_properties.json is in the .vscode in my code folder, but there is no hint for its existing

Also, I have no idea why this "configurationProvider": "ms-vscode.makefile-tools" happen to me, I am new to VS Code, and definitely don't know what's the purpose of configurationProvider. I just installed some extensions according to VS Code suggestion or Google suggestion..

**Reply by sean-mcmanus** [reply]
Yeah, it looks like the Makefile Tool extension may not have been configured correctly or it could be a bug with that extension, since it was returning an empty browse path, causing no workspace symbols to be parsed.

We may have automatically set the configuration provider when we detected a makefile. The purpose of the configuration provider is to provide better IntelliSense configuration than what C_Cpp.default.* or c_cpp_properties.json could provide. Using compileCommands is another alternative for configuration too.
